src/query/plan/operator.cpp

```diff
@@ -3007,58 +3007,71 @@ bool Delete::DeleteCursor::Pull(Frame &frame, ExecutionContext &context) {
   OOMExceptionEnabler oom_exception;
   SCOPED_PROFILE_OP("Delete");
 
-  if (delete_executed_) {
-    return false;
+  if (self_.buffer_size_ && buffer_size_ == -1) {
+    // Limit expression doesn't contain identifiers so graph view is not
+    // important.
+    ExpressionEvaluator evaluator(&frame, context.symbol_table, context.evaluation_context, context.db_accessor,
+                                  storage::View::OLD);
+    TypedValue buffer_size = self_.buffer_size_->Accept(evaluator);
+    if (buffer_size.type() != TypedValue::Type::Int)
+      throw QueryRuntimeException("Number of periodically deleted elements must be an integer.");
+
+    buffer_size_ = buffer_size.ValueInt();
+    if (buffer_size_ < 0) throw QueryRuntimeException("Number of periodically deleted elements must be non-negative.");
   }
 
-  if (input_cursor_->Pull(frame, context)) {
+  bool const has_more = input_cursor_->Pull(frame, context);
+
+  if (has_more) {
     UpdateDeleteBuffer(frame, context);
-    return true;
+    pulled_++;
   }
 
-  auto &dba = *context.db_accessor;
-  auto res = dba.DetachDelete(std::move(buffer_.nodes), std::move(buffer_.edges), self_.detach_);
-  if (res.HasError()) {
-    switch (res.GetError()) {
-      case storage::Error::SERIALIZATION_ERROR:
-        throw TransactionSerializationException();
-      case storage::Error::VERTEX_HAS_EDGES:
-        throw RemoveAttachedVertexException();
-      case storage::Error::DELETED_OBJECT:
-      case storage::Error::PROPERTIES_DISABLED:
-      case storage::Error::NONEXISTENT_OBJECT:
-        throw QueryRuntimeException("Unexpected error when deleting a node.");
+  if (!has_more || (buffer_size_ > -1 && pulled_ >= buffer_size_)) {
```

Comment:
It would be nicer if buffer_size is optional here, so you can have something like
```
if (!has_more || (buffer_size.has_value() && ...)
...
```